librz/bin/pdb/pdb.c

```diff
@@ -1,1433 +1,559 @@
-	rz_list_foreach (types, it, type) {
-		STypeInfo *type_info = &type->type_data;
-		// skip unprintable types and enums
-		if (!type || !is_printable_type(type_info->leaf_type) || type_info->leaf_type == eLF_ENUM) {
-			continue;
-		}
-		// skip forward references
-		if (type_info->is_fwdref) {
-			if (type_info->is_fwdref(type_info)) {
-				continue;
-			}
-		}
-		char *name = NULL;
-		if (type_info->get_name) {
-			name = type_info->get_name(type_info);
-		}
-		if (!name) {
-			name = create_type_name_from_offset(type->tpi_idx);
-			to_free_name = true;
-		}
-		RzList *members = NULL;
-		if (type_info->get_members) {
-			members = type_info->get_members(type_info);
-			if (!members) {
-				continue;
-			}
-		}
-		// pf.name <format chars> <member names>
-		RzStrBuf format;
-		rz_strbuf_init(&format);
-		RzStrBuf member_names;
-		rz_strbuf_init(&member_names);
-
-		if (type_info->leaf_type == eLF_UNION) {
-			rz_strbuf_append(&format, "0"); // every type start from the offset 0
-		}
-
-		RzListIter *member_iter;
-		STypeInfo *member_info;
-		rz_list_foreach (members, member_iter, member_info) {
-			switch (type_info->leaf_type) {
-			case eLF_STRUCTURE_19:
-			case eLF_CLASS_19:
-			case eLF_STRUCTURE:
-			case eLF_CLASS:
-			case eLF_UNION:
-				if (build_member_format(member_info, &format, &member_names) == -1) { // if failed
-					goto fail; // skip to the next one, we can't build format from this
-				}
-				break;
-			default:
-				rz_warn_if_reached();
-			}
-			rz_strbuf_append(&member_names, " ");
-		}
-		if (format.len == 0) { // if it has no members, it's useless for us then?
-			goto fail;
-		}
-		char *sanitized_name = rz_str_sanitize_sdb_key(name);
-		pdb->cb_printf("pf.%s %s %s\n", sanitized_name, rz_strbuf_get(&format), rz_strbuf_get(&member_names));
-
-		if (to_free_name) { // name can be generated or part of the PDB data
-			RZ_FREE(name);
-		}
-		RZ_FREE(sanitized_name);
-		rz_strbuf_fini(&format);
-		rz_strbuf_fini(&member_names);
-
-	fail: // if we can't print whole type correctly, don't print at all
-		if (to_free_name) {
-			RZ_FREE(name);
-		}
-		rz_strbuf_fini(&format);
-		rz_strbuf_fini(&member_names);
+		pj_end(pj);
 	}
 }
 
 /**
  * \brief Prints out all the type information in regular,json or pf format
- * 
+ *
  * \param pdb PDB information
  * \param mode printing mode
  */
-static void print_types(const RzPdb *pdb, PJ *pj, const int mode) {
-	RzList *plist = pdb->pdb_streams;
-	STpiStream *tpi_stream = rz_list_get_n(plist, ePDB_STREAM_TPI);
+RZ_API void rz_bin_pdb_print_types(RzTypeDB *db, const RzPdb *pdb, PJ *pj,
+	const int mode) {
+	TpiStream *stream = pdb->s_tpi;
 
-	if (!tpi_stream) {
+	if (!stream) {
 		eprintf("There is no tpi stream in current pdb\n");
 		return;
 	}
 	switch (mode) {
-	case 'd': print_types_regular(pdb, tpi_stream->types); return;
-	case 'j': print_types_json(pdb, pj, tpi_stream->types); return;
-	case 'r': print_types_format(pdb, tpi_stream->types); return;
+	case 'd':
```

Comment:
Since you edit this function anyway, I recommend to switch to `RzOutputMode` enum, see https://github.com/rizinorg/rizin/issues/489